docs/current_docs/developer-guide/go/742930-services.mdx

```diff
@@ -0,0 +1,260 @@
+```
+
+Here is an example call for this Dagger Function:
+
+```shell
+dagger call user-list --host-service=tcp://localhost:3306
+```
+
+The result will be:
+
+```shell
+Host    User
+%       root
+localhost       mariadb.sys
+localhost       root
+```
+
+## Persist service state
+
+Dagger cancels each service run after a 10 second grace period to avoid frequent restarts. To avoid relying on the grace period, use a cache volume to persist a service's data, as in the following example:
+
+```go file=./snippets/services/persist-service-state/main.go
+```
+
+This function uses Redis's SAVE command to ensure data is synced. By default, Redis flushes data to disk periodically.
+
+Here is an example call for this Dagger Function:
+
+```shell
+dagger call redis-service
+```
+
+The result will be:
+
+```shell
+abc
+```
+
+## Start and stop services
+
+Services are designed to be expressed as a Directed Acyclic Graph (DAG) with explicit bindings allowing services to be started lazily, just like every other DAG node. But sometimes, you may need to explicitly manage the lifecycle in a Dagger Function.
+
+Here's an example which demonstrates explicitly starting a Docker daemon for use in a test suite:
+
+```go file=./snippets/services/start-stop-services/main.go
+```
+
+## Example: MariaDB database service for application tests
+
+The following example demonstrates how services can be used in Dagger Functions, by creating a Dagger Function for application unit/integration testing against a bound MariaDB database service.
+
+The application used in this example is [Drupal](https://www.drupal.org/), a popular open-source PHP CMS. Drupal includes a large number of unit tests, including tests which require an active database connection. All Drupal 10.x tests are written and executed using the [PHPUnit](https://phpunit.de/) testing framework. Read more about [running PHPUnit tests in Drupal](https://phpunit.de/).
+
+```go file=./snippets/services/test-against-db-service/main.go
+```
+
+Here is an example call for this Dagger Function:
+
+```shell
+dagger call test
+```
+
+The result will be:
+
+```shell
+PHPUnit 9.6.17 by Sebastian Bergmann and contributors.
+Runtime:       PHP 8.2.5
+Configuration: /opt/drupal/web/core/phpunit.xml.dist
+Testing
+.....................S                                            22 / 22 (100%)
+Time: 00:15.806, Memory: 315.00 MB
+There was 1 skipped test:
+
+1) Drupal\Tests\pgsql\Kernel\pgsql\KernelTestBaseTest::testSetUp
+
+This test only runs for the database driver 'pgsql'. Current database driver is 'mysql'.
+/opt/drupal/web/core/tests/Drupal/KernelTests/Core/Database/DriverSpecificKernelTestBase.php:44
+/opt/drupal/vendor/phpunit/phpunit/src/Framework/TestResult.php:728
+
+OK, but incomplete, skipped, or risky tests!
+Tests: 22, Assertions: 72, Skipped: 1.
+```
+
+## Reference: How service binding works in Dagger Functions
```

Comment:
This entire section and snippets should be re-checked and may need changes. @vito lmk if you prefer to omit it atm to avoid blocking release of this page.